app/controllers/compositions_controller.rb

```diff
@@ -50,21 +51,22 @@ def get_map_segment_ids(heroes, composition)
     end
 
     player_selections = composition.player_selections.
-      includes(player_hero: :player)
+      includes(:player)
     player_selections.each do |player_selection|
-      hero_id = player_selection.player_hero.hero_id
-      player_name = player_selection.player_hero.player.name
-      result[hero_id][player_name] = player_selection.map_segment_id
+      hero_id = player_selection.hero_id
+      player_name = player_selection.player.name
+      result[hero_id][player_name] ||= []
+      result[hero_id][player_name] << player_selection.map_segment_id
     end
 
     result
   end
```

Comment:
> If there's another reason we're requiring all Heroes to be populated first in the result hash

For the convenience of the JSON views, so we don't have much logic in the views. When I tried your code, I got a `undefined method '[]' for nil:NilClass` error on `last_composition.json.jbuilder:25` I believe because `@map_segment_ids[hero.id]` is nil so we can't grab `@map_segment_ids[hero.id][player.name]` from it.